### PR TITLE
Update more acquisition function builders

### PR DESCRIPTION
**Pull request by uri-granta**
Update the following acquisition function builders to support in place updates and avoid retracing:

* ExpectedConstrainedImprovement
* ExpectedHypervolumeImprovement
* ExpectedConstrainedHypervolumeImprovement

Still to do (in a separate PR):

* BatchMonteCarloExpectedHypervolumeImprovement

### CODE REVIEW

File: trieste/acquisition/function.py [hunk 1]
```diff
@@ -706,6 +706,9 @@ def __init__(
         self._objective_tag = objective_tag
         self._constraint_builder = constraint_builder
         self._min_feasibility_probability = min_feasibility_probability
+        self._constraint_fn: Optional[AcquisitionFunction] = None
+        self._improvement_fn: Optional[AcquisitionFunction] = None
+        self._constrained_fn: Optional[AcquisitionFunction] = None
```
Comment (apaleyes) [inline trieste/acquisition/function.py]: Can these have a bit more distinctive names? It is quite hard to figure out which of two `_constrain...` is being used while reading the code.

File: trieste/acquisition/function.py [hunk 2]
```diff
@@ -735,7 +738,52 @@ def prepare_acquisition_function(
             " objective data, but the objective data is empty.",
         )
 
-        constraint_fn = self._constraint_builder.prepare_acquisition_function(datasets, models)
+        self._constraint_fn = self._constraint_builder.prepare_acquisition_function(
+            datasets, models
+        )
+        pof = self._constraint_fn(objective_dataset.query_points[:, None, ...])
```
Comment (apaleyes) [inline trieste/acquisition/function.py:744]: perhaps it won't hurt to spell out `pof` here?

File: trieste/acquisition/function.py [hunk 3]
```diff
@@ -745,9 +793,20 @@ def prepare_acquisition_function(
         feasible_query_points = tf.boolean_mask(objective_dataset.query_points, is_feasible)
         feasible_mean, _ = objective_model.predict(feasible_query_points)
         eta = tf.reduce_min(feasible_mean, axis=0)
-        function = expected_improvement(objective_model, eta)
+        tf.debugging.Assert(isinstance(self._improvement_fn, expected_improvement), [])
+        self._improvement_fn.update(eta)  # type: ignore
```
Comment (apaleyes) [inline trieste/acquisition/function.py]: Is the expectation that if lines 747-748 are entered (early return in `prepare`), execution never reaches this line here? If it does, this assertion will fail

File: trieste/acquisition/function.py [hunk 4]
```diff
@@ -735,7 +738,52 @@ def prepare_acquisition_function(
...
+        eta = tf.reduce_min(feasible_mean, axis=0)
+        self._improvement_fn = expected_improvement(objective_model, eta)
+
+        @tf.function
+        def constrained_function(x: TensorType) -> TensorType:
+            return cast(AcquisitionFunction, self._improvement_fn)(x) * cast(
+                AcquisitionFunction, self._constraint_fn
+            )(x)
+
+        self._constrained_fn = cast(AcquisitionFunction, constrained_function)
```
Comment (apaleyes) [inline trieste/acquisition/function.py]: Why does this cast work?

File: trieste/acquisition/function.py [hunk 5]
```diff
@@ -745,9 +793,20 @@ def prepare_acquisition_function(
...
+            return self._constrained_fn
+
+        @tf.function
+        def constrained_function(x: TensorType) -> TensorType:
+            return cast(AcquisitionFunction, self._improvement_fn)(x) * cast(
+                AcquisitionFunction, self._constraint_fn
+            )(x)
 
-        return lambda at: function(at) * constraint_fn(at)
+        self._constrained_fn = constrained_function
```
Comment (apaleyes) [inline trieste/acquisition/function.py]: There is a cast in similar situation in `prepare`

File: trieste/acquisition/function.py [hunk 6]
```diff
@@ -735,7 +738,52 @@ def prepare_acquisition_function(
...
+        tf.debugging.assert_positive(
+            len(objective_dataset),
+            message="Expected improvement is defined with respect to existing points in the"
+            " objective data, but the objective data is empty.",
+        )
+        tf.debugging.Assert(None not in [self._constraint_fn], [])
+
+        constraint_fn = cast(AcquisitionFunction, self._constraint_fn)
+        self._constraint_builder.update_acquisition_function(constraint_fn, datasets, models)
         pof = constraint_fn(objective_dataset.query_points[:, None, ...])
```
Comment (apaleyes) [inline trieste/acquisition/function.py:786]: ditto about `pof`

File: trieste/acquisition/function.py (hunk 1 above)
Comment (uri-granta) [reply trieste/acquisition/function.py]: :+1: 

FYI this PR isn't actually ready yet (I should have put it in draft, sorry). Still missing new tests, and I'm panning to convert another few functions as part of the same PR.

File: trieste/acquisition/function.py (hunk 3 above)
Comment (uri-granta) [reply trieste/acquisition/function.py]: Good catch. This does need to be handled (and I'll make sure it's covered in a unit test).

File: trieste/acquisition/function.py (hunk 4 above)
Comment (uri-granta) [reply trieste/acquisition/function.py]: This cast shouldn't actually be necessary, as mypy should know that constrained_function is an AcquisitionFunction and not None. For some reason though it immediately forgets this on the next line. Though it'd probably be clearer to remove the cast and just return constrained_function instead.

File: trieste/acquisition/function.py (hunk 5 above)
Comment (uri-granta) [reply trieste/acquisition/function.py]: See comment above. The reason a cast isn't necessary here is that mypy infers that self._constrained_fn can't be None due to the earlier if clause. (Though as I said, I'll remove the cast from further up too.)

File: trieste/acquisition/function.py (hunk 1 above)
Comment (apaleyes) [reply trieste/acquisition/function.py]: ok, noted, will hold off further reviewing for a bit

File: trieste/acquisition/function.py [hunk 7]
```diff
@@ -744,10 +791,36 @@ def prepare_acquisition_function(
...
+        """
         eta = tf.reduce_min(feasible_mean, axis=0)
-        function = expected_improvement(objective_model, eta)
 
-        return lambda at: function(at) * constraint_fn(at)
+        if self._expected_improvement_fn is None:
+            self._expected_improvement_fn = expected_improvement(objective_model, eta)
+        else:
+            tf.debugging.Assert(isinstance(self._expected_improvement_fn, expected_improvement), [])
+            self._expected_improvement_fn.update(eta)  # type: ignore
```
Comment (henrymoss) [inline trieste/acquisition/function.py:823]: It there any reason why this is its own function?

Comment (uri-granta) [reply trieste/acquisition/function.py:823]: Yes, it allows ExpectedConstrainedHypervolumeImprovement to inherit from ExpectedConstrainedImprovement without having to duplicate lots of code. Also, it's now used both when preparing and updating the functions.

File: trieste/acquisition/function.py [hunk 8]
```diff
@@ -779,37 +852,70 @@ def prepare_acquisition_function(
...
+            partition_bounds[1], trainable=False, shape=[None, partition_bounds[1].shape[-1]]
+        )
+        self._cross_index = tf.constant(
+            list(product(*[[0, 1]] * self._lb_points.shape[-1]))
+        )  # [2^d, indices_at_dim]
+
+    def update(self, partition_bounds: tuple[TensorType, TensorType]) -> None:
+        """Update the acquisition function with new partition bounds."""
+        self._lb_points.assign(partition_bounds[0])
+        self._ub_points.assign(partition_bounds[1])
```
Comment (apaleyes) [inline trieste/acquisition/function.py:915]: May we ever want to serialize an acquisition object? If there is a chance, e.g. pickle the whole BayesianOptimizer object for some reason, it's probably worth ensuring dynamic shapes here, as we did in models recently. If serializing acquisition object makes no sense, don't bother

File: trieste/acquisition/function.py [hunk 9]
```diff
@@ -779,37 +852,70 @@ def prepare_acquisition_function(
...
 
+    def update_acquisition_function(
+        self, function: AcquisitionFunction, dataset: Dataset, model: ProbabilisticModel
+    ) -> AcquisitionFunction:
+        """
+        :param function: The acquisition function to update.
+        :param dataset: The data from the observer.
+        :param model: The model over the specified ``dataset``.
+        """
+        tf.debugging.assert_positive(len(dataset), message="Dataset must be populated.")
```
Comment (apaleyes) [inline trieste/acquisition/function.py:863]: For education more than anything else: why use TF's assert instead of just plain if...raise?

File: trieste/acquisition/function.py [hunk 10]
```diff
@@ -735,7 +738,51 @@ def prepare_acquisition_function(
...
+        """
+        objective_model = models[self._objective_tag]
+        objective_dataset = datasets[self._objective_tag]
+
+        tf.debugging.assert_positive(
+            len(objective_dataset),
+            message="Expected improvement is defined with respect to existing points in the"
+            " objective data, but the objective data is empty.",
+        )
+        tf.debugging.Assert(None not in [self._constraint_fn], [])
```
Comment (apaleyes) [inline trieste/acquisition/function.py]: Is that just `self._constraint_fn is not None`?

File: tests/unit/acquisition/test_function.py [hunk 11]
```diff
@@ -1692,18 +1738,29 @@ def prepare_acquisition_function(
         ) -> AcquisitionFunction:
             return lambda x: tf.ones_like(tf.squeeze(x, -2))
 
-    data = {"foo": Dataset(train_x, model_pred_observation)}
+    data = {"foo": Dataset(train_x[:5], model_pred_observation[:5])}
```
Comment (apaleyes) [inline tests/unit/acquisition/test_function.py:1741]: Why are these sliced now?

File: tests/unit/acquisition/test_function.py [hunk 12]
```diff
@@ -1692,18 +1738,29 @@ def prepare_acquisition_function(
...
     npt.assert_allclose(echvi(at), ehvi(at))
 
+    new_data = {"foo": Dataset(train_x, model_pred_observation)}
+    up_echvi = builder.update_acquisition_function(echvi, new_data, models_)
+    assert up_echvi == echvi
+    up_ehvi = (
+        ExpectedHypervolumeImprovement()
+        .using("foo")
+        .prepare_acquisition_function(new_data, models_)
+    )
```
Comment (apaleyes) [inline tests/unit/acquisition/test_function.py:1759]: Why is that necessary, if basically the same happened two lines above?

File: trieste/acquisition/function.py [hunk 13]
```diff
@@ -779,37 +852,70 @@ def prepare_acquisition_function(
...
+        """
+        self._model = model
+        self._lb_points = tf.Variable(
+            partition_bounds[0], trainable=False, shape=[None, partition_bounds[0].shape[-1]]
+        )
+        self._ub_points = tf.Variable(
+            partition_bounds[1], trainable=False, shape=[None, partition_bounds[1].shape[-1]]
+        )
+        self._cross_index = tf.constant(
+            list(product(*[[0, 1]] * self._lb_points.shape[-1]))
```
Comment (apaleyes) [inline trieste/acquisition/function.py:909]: Not sure I understand what cross_index is for, probably worth explaining in a comment. Also:

```suggestion
            list(product([0, 1] * self._lb_points.shape[-1]))
```

**Review by apaleyes — COMMENTED** [review]
Nothing major, just some nitpicking and clarifying questions

File: trieste/acquisition/function.py (hunk 9 above)
Comment (uri-granta) [reply trieste/acquisition/function.py:863]: If ... raise doesn't work well with functions wrapped with tf.function.

File: trieste/acquisition/function.py (hunk 10 above)
Comment (uri-granta) [reply trieste/acquisition/function.py]: I'm pretty sure there was some obscure tensorflow reason, but changing it doesn't seem to make the unit tests fail! I'll update this (and a few other similar cases) and see if anything breaks.

File: tests/unit/acquisition/test_function.py (hunk 11 above)
Comment (uri-granta) [reply tests/unit/acquisition/test_function.py:1741]: The testnow  prepares the function with some of the data, then later updates it with all of the data (to test that updating works without retracing).

File: tests/unit/acquisition/test_function.py (hunk 12 above)
Comment (uri-granta) [reply tests/unit/acquisition/test_function.py:1759]: This is creating an updated ehvi function based on the new data to compare the updated echvi function (it's actually creating a brand new function but that's ok as we're not testing updating ehvi here, only echvi).

File: trieste/acquisition/function.py (hunk 13 above)
Comment (uri-granta) [reply trieste/acquisition/function.py:909]: Not quite. Product takes a list of arguments, not a single argument that's a list: `product([0,1], [0,1], [0,1])` not `product([[0,1],[0,1],[0,1]])` or `product([0,1,0,1,0,1])`.

Comment (apaleyes) [reply trieste/acquisition/function.py:909]: i stand corrected, thanks

File: tests/unit/acquisition/test_function.py (hunk 12 above)
Comment (apaleyes) [reply tests/unit/acquisition/test_function.py:1759]: ok, i see , thanks

File: tests/unit/acquisition/test_function.py (hunk 11 above)
Comment (apaleyes) [reply tests/unit/acquisition/test_function.py:1741]: got it

File: trieste/acquisition/function.py (hunk 8 above)
Comment (uri-granta) [reply trieste/acquisition/function.py:915]: I can't immediately see why we'd do this, so will leave it for now.

Comment (apaleyes) [reply trieste/acquisition/function.py:915]: ok!

**Review by apaleyes — APPROVED** [review]

**Review by henrymoss — APPROVED** [review]
